torch/_subclasses/fake_tensor.py

```diff
@@ -2589,3 +2599,66 @@ def dump_cache_stats() -> None:
         width = max(len(k) for k in bypasses)
         for k, v in sorted(bypasses.items(), key=lambda i: -i[1]):
             log.info("    %-*s %s", width + 1, f"{k}:", v)
+
+
+def inferred_fake_kernel_from_real_out(
+    mode: FakeTensorMode, op: torch._ops.OpOverload, real_out: Any
+) -> Any:
+    assert mode.shape_env is not None
+
+    # Only support operators that have all Tensor outputs
+    real_flat_out, spec = pytree.tree_flatten(real_out)
+    if not all(isinstance(t, torch.Tensor) for t in real_flat_out):
+        raise RuntimeError(
+            f"propagate_real_tensors: we don't support operators that return "
+            f"non-Tensors. Got {op._name}"
+        )
+
+    def make_fake(real_out: torch.Tensor) -> torch.Tensor:
+        def unsupported(reason: str) -> None:
+            raise RuntimeError(
+                f"propagate_real_tensors: we cannot infer a Fake kernel "
+                f"(meta kernel) for operator {op._name} because {reason}. "
+                f"Please use torch.library.register_fake to add a Fake kernel."
+            )
+
+        if real_out.storage_offset() != 0:
+            unsupported(
+                f"a return has a non-zero storage offset {real_out.storage_offset()}"
+            )
+
+        # Assume rank specialized. That is, none of the inputs will affect the rank.
+        # This is emperically almost always true. An example of something that
+        # isn't is an `unsqueeze(x, n_times)` operator.
+        # A better thing to do is to *also* insert a runtime assertion for the rank of
+        # the output.
+        fake_shape = [
+            torch._library.fake_impl.allocate_size(mode.shape_env)
+            for _ in range(real_out.dim())
+        ]
+        fake_strides = [-1] * real_out.dim()
+
+        strides = [(s, idx) for idx, s in enumerate(real_out.stride())]
+        strides.sort()
+        expected = 1
+        fake_stride = expected
+        for s, idx in strides:
+            if s != expected:
+                unsupported(
+                    f"a return was not dense in memory (sizes {real_out.shape} strides {real_out.stride()})"
+                )
+            fake_strides[idx] = fake_stride
+            expected = expected * real_out.shape[idx]
+            fake_stride = fake_stride * fake_shape[idx]
```

Comment:
just for my understanding, why do we do need to have the `.sort()` call above, and why can't we just return the real tensor strides as-is, even if they're not dense in memory?

Reply:
>  why do we do need to have the .sort() call above

We're using this to check the tensor is dense in memory. If the tensor has sizes (a, b, c, d), then it is contiguous if the strides look like (b * c * d, c * d, d, 1). The tensor is dense in memory if the tensor can be permuted into being contiguous. We sort the strides to check if they follow that pattern

Reply:
this makes sense, thank you!